File: py/tests/local/test_activepieces.py
"""
Full ActivePieces resource test — sync + async, runs against prodv2 gateway.
Usage:
    pip install -e ".[dev]"
    python tests/local/test_activepieces.py
"""
import os
import asyncio
from imbrace import ImbraceClient
from imbrace.async_client import AsyncImbraceClient

ACCESS_TOKEN = os.environ.get("IMBRACE_ACCESS_TOKEN", "acc_c8c27f3b-e147-4735-b641-61e8d3706692")
GATEWAY = os.environ.get("IMBRACE_GATEWAY_URL", "https://app-gatewayv2.imbrace.co")

passed = 0
failed = 0
skipped = 0
created = {"flow_id": None, "folder_id": None}


def ok(label, detail=""):
    global passed
    detail_str = f"  →  {str(detail)[:80]}" if detail else ""
    print(f"  ✓ {label}{detail_str}")
    passed += 1


def fail(label, err):
    global failed
    print(f"  ✗ {label}: {err}")
    failed += 1


def skip(label, reason):
    global skipped
    print(f"  - {label}  (skipped: {reason})")
    skipped += 1


# ── Sync Tests ────────────────────────────────────────────────────────────────

def test_sync():
    global created
    client = ImbraceClient(access_token=ACCESS_TOKEN, gateway=GATEWAY)
    ap = client.activepieces

    print("\n[1] Flows (sync)")
    try:
        res = ap.list_flows(limit=3)
        ok("list_flows()", f"{len(res['data'])} flows")
        if res["data"]:
            flow = ap.get_flow(res["data"][0]["id"])
            ok("get_flow()", f"id={flow['id']} status={flow['status']}")
            project_id = res["data"][0]["projectId"]
        else:
            skip("get_flow()", "no flows")
            project_id = None
    except Exception as e:
        fail("Flows", e)
        project_id = None

    print("\n[2] Create Flow (sync)")
    try:
        if not project_id:
            skip("create_flow()", "no projectId")
        else:
            flow = ap.create_flow("SDK Python Test Flow", project_id)
            created["flow_id"] = flow["id"]
            ok("create_flow()", f"id={flow['id']}")
    except Exception as e:
        fail("create_flow()", e)

    print("\n[3] Trigger Flow (sync)")
    try:
        flows = ap.list_flows(limit=3)
        enabled = next((f for f in flows["data"] if f["status"] == "ENABLED"), None)
        if not enabled:
            skip("trigger_flow()", "no ENABLED flow")
        else:
            res = ap.trigger_flow(enabled["id"], {"test": True, "source": "sdk-python-test"})
            ok("trigger_flow()", str(res)[:60])
    except Exception as e:
        fail("trigger_flow()", e)

    print("\n[4] Flow Runs (sync)")
    try:
        res = ap.list_runs(limit=3)
        ok("list_runs()", f"{len(res['data'])} runs")
        if res["data"]:
            run = ap.get_run(res["data"][0]["id"])
            ok("get_run()", f"id={run['id']} status={run['status']}")
        else:
            skip("get_run()", "no runs")
    except Exception as e:
        fail("Flow Runs", e)

    print("\n[5] Folders (sync)")
    try:
        if not project_id:
            skip("Folders", "no projectId")
        else:
            res = ap.list_folders(limit=3)
            ok("list_folders()", f"{len(res['data'])} folders")

            folder = ap.create_folder("SDK Python Test Folder", project_id)
            created["folder_id"] = folder["id"]
            ok("create_folder()", f"id={folder['id']}")

            updated = ap.update_folder(folder["id"], "SDK Python Test Folder Updated")
            ok("update_folder()", f"name={updated['displayName']}")

            fetched = ap.get_folder(folder["id"])
            ok("get_folder()", f"id={fetched['id']}")
    except Exception as e:
        fail("Folders", e)

    print("\n[6] App Connections (sync)")
    try:
        res = ap.list_connections(limit=3)
        ok("list_connections()", f"{len(res['data'])} connections")
        if res["data"]:
            conn = ap.get_connection(res["data"][0]["id"])
            ok("get_connection()", f"id={conn['id']}")
        else:
            skip("get_connection()", "no connections")
    except Exception as e:
        fail("App Connections", e)

    print("\n[7] Pieces (sync)")
    try:
        res = ap.list_pieces(limit=5)
        ok("list_pieces()", f"{len(res)} pieces, first={res[0]['displayName'] if res else 'n/a'}")
    except Exception as e:
        fail("list_pieces()", e)

    print("\n[8] Triggers (sync)")
    try:
        res = ap.get_trigger_run_status()
        ok("get_trigger_run_status()", f"{len(res.get('pieces', {}))} pieces tracked")
    except Exception as e:
        fail("get_trigger_run_status()", e)

    print("\n[9] Tables & Records (sync)")
    try:
        tables = ap.list_tables(limit=3)
        ok("list_tables()", f"{len(tables['data'])} tables")
        if tables["data"]:
            table = ap.get_table(tables["data"][0]["id"])
            ok("get_table()", f"id={table['id']}")
            records = ap.list_records(table["id"], limit=3)
            ok("list_records()", f"{len(records['data'])} records")
        else:
            skip("get_table() / list_records()", "no tables")
    except Exception as e:
        fail("Tables & Records", e)

    print("\n[10] MCP Servers (sync)")
    try:
        if not project_id:
            skip("MCP Servers", "no projectId")
        else:
            res = ap.list_mcp_servers(project_id)
            ok("list_mcp_servers()", f"{len(res['data'])} MCP servers")
            if res["data"]:
                server = ap.get_mcp_server(res["data"][0]["id"])
                ok("get_mcp_server()", f"id={server['id']}")
            else:
                skip("get_mcp_server()", "no MCP servers")
    except Exception as e:
        fail("MCP Servers", e)

    print("\n[11] User Invitations (sync)")
    try:
        res = ap.list_invitations(type="PLATFORM", limit=3)
        ok("list_invitations(PLATFORM)", f"{len(res['data'])} invitations")
    except Exception as e:
        fail("list_invitations(PLATFORM)", e)
    try:
        res = ap.list_invitations(type="PROJECT", limit=3)
        ok("list_invitations(PROJECT)", f"{len(res['data'])} invitations")
    except Exception as e:
        fail("list_invitations(PROJECT)", e)

    print("\n[12] Cleanup (sync)")
    if created["folder_id"]:
        try:
            ap.delete_folder(created["folder_id"])
            ok(f"delete_folder({created['folder_id']})")
        except Exception as e:
            fail("delete_folder()", e)
    if created["flow_id"]:
        try:
            ap.delete_flow(created["flow_id"])
            ok(f"delete_flow({created['flow_id']})")
        except Exception as e:
            fail("delete_flow()", e)


# ── Async Tests ───────────────────────────────────────────────────────────────

async def test_async():
    created_async = {"flow_id": None, "folder_id": None}
    client = AsyncImbraceClient(access_token=ACCESS_TOKEN, gateway=GATEWAY)
    ap = client.activepieces

    print("\n[13] Flows (async)")
    try:
        res = await ap.list_flows(limit=2)
        ok("async list_flows()", f"{len(res['data'])} flows")
        project_id = res["data"][0]["projectId"] if res["data"] else None
    except Exception as e:
        fail("async Flows", e)
        project_id = None

    print("\n[14] Create + Delete Flow (async)")
    try:
        if not project_id:
            skip("async create_flow()", "no projectId")
        else:
            flow = await ap.create_flow("SDK Python Async Test Flow", project_id)
            created_async["flow_id"] = flow["id"]
            ok("async create_flow()", f"id={flow['id']}")
    except Exception as e:
        fail("async create_flow()", e)

    print("\n[15] Folders (async)")
    try:
        if not project_id:
            skip("async Folders", "no projectId")
        else:
            res = await ap.list_folders(limit=2)
            ok("async list_folders()", f"{len(res['data'])} folders")

            folder = await ap.create_folder("SDK Python Async Folder", project_id)
            created_async["folder_id"] = folder["id"]
            ok("async create_folder()", f"id={folder['id']}")
    except Exception as e:
        fail("async Folders", e)

    print("\n[16] Runs + Pieces + Triggers (async)")
    try:
        runs = await ap.list_runs(limit=2)
        ok("async list_runs()", f"{len(runs['data'])} runs")
        pieces = await ap.list_pieces(limit=3)
        ok("async list_pieces()", f"{len(pieces)} pieces")
        status = await ap.get_trigger_run_status()
        ok("async get_trigger_run_status()", f"{len(status.get('pieces', {}))} pieces")
    except Exception as e:
        fail("async Runs/Pieces/Triggers", e)

    print("\n[17] Async Cleanup")
    if created_async["folder_id"]:
        try:
            await ap.delete_folder(created_async["folder_id"])
            ok(f"async delete_folder({created_async['folder_id']})")
        except Exception as e:
            fail("async delete_folder()", e)
    if created_async["flow_id"]:
        try:
            await ap.delete_flow(created_async["flow_id"])
            ok(f"async delete_flow({created_async['flow_id']})")
        except Exception as e:
            fail("async delete_flow()", e)

    await client.close()


# ── Main ──────────────────────────────────────────────────────────────────────

if __name__ == "__main__":
    test_sync()
    asyncio.run(test_async())

    print(f"\n{'─' * 55}")
    print(f"  {passed} passed  |  {failed} failed  |  {skipped} skipped")
    if failed > 0:
        raise SystemExit(1)

File: ts/tests/local/test-activepieces.mjs
/**
 * Full ActivePieces resource test — runs against prodv2 gateway.
 * Usage: node test-activepieces.mjs
 */

import { ImbraceClient } from '@imbrace/sdk'

const client = new ImbraceClient({
  accessToken: process.env.IMBRACE_ACCESS_TOKEN || 'acc_c8c27f3b-e147-4735-b641-61e8d3706692',
  gateway: process.env.IMBRACE_GATEWAY_URL || 'https://app-gatewayv2.imbrace.co',
})

const ap = client.activepieces

let passed = 0
let failed = 0
let skipped = 0

// track created resources for cleanup
const created = { flowId: null, folderId: null, connectionId: null, invitationId: null }

function ok(label, detail = '') {
  console.log(`  ✓ ${label}${detail ? `  →  ${String(detail).slice(0, 80)}` : ''}`)
  passed++
}

function fail(label, err) {
  console.error(`  ✗ ${label}: ${err?.message ?? JSON.stringify(err)}`)
  failed++
}

function skip(label, reason) {
  console.log(`  - ${label}  (skipped: ${reason})`)
  skipped++
}

// ── Flows ────────────────────────────────────────────────────────────────────

console.log('\n[1] Flows')
try {
  const res = await ap.listFlows({ limit: 3 })
  ok('listFlows()', `${res.data.length} flows, next=${res.next}`)

  if (res.data.length > 0) {
    const flow = await ap.getFlow(res.data[0].id)
    ok('getFlow()', `id=${flow.id} status=${flow.status}`)
  } else {
    skip('getFlow()', 'no flows available')
  }
} catch (e) { fail('Flows', e) }

// ── Create Flow ──────────────────────────────────────────────────────────────

console.log('\n[2] Create Flow')
try {
  // Need a projectId — get it from existing flow
  const flows = await ap.listFlows({ limit: 1 })
  const projectId = flows.data[0]?.projectId
  if (!projectId) throw new Error('No projectId available from existing flows')

  const flow = await ap.createFlow({ displayName: 'SDK Test Flow', projectId })
  created.flowId = flow.id
  ok('createFlow()', `id=${flow.id} name=${flow.version?.displayName}`)
} catch (e) { fail('createFlow()', e) }

// ── Trigger Flow ──────────────────────────────────────────────────────────────

console.log('\n[3] Trigger Flow (webhook)')
try {
  const flows = await ap.listFlows({ limit: 3 })
  const enabledFlow = flows.data.find(f => f.status === 'ENABLED')
  if (!enabledFlow) {
    skip('triggerFlow()', 'no ENABLED flow found')
  } else {
    const res = await ap.triggerFlow(enabledFlow.id, { test: true, source: 'sdk-test' })
    ok('triggerFlow()', `response=${JSON.stringify(res).slice(0, 60)}`)
  }
} catch (e) { fail('triggerFlow()', e) }

// ── Flow Runs ────────────────────────────────────────────────────────────────

console.log('\n[4] Flow Runs')
try {
  const res = await ap.listRuns({ limit: 3 })
  ok('listRuns()', `${res.data.length} runs`)

  if (res.data.length > 0) {
    const run = await ap.getRun(res.data[0].id)
    ok('getRun()', `id=${run.id} status=${run.status}`)
  } else {
    skip('getRun()', 'no runs available')
  }
} catch (e) { fail('Flow Runs', e) }

// ── Folders ───────────────────────────────────────────────────────────────────

console.log('\n[5] Folders')
try {
  const flows = await ap.listFlows({ limit: 1 })
  const projectId = flows.data[0]?.projectId
  if (!projectId) throw new Error('No projectId available')

  const list = await ap.listFolders({ limit: 3 })
  ok('listFolders()', `${list.data.length} folders`)

  const folder = await ap.createFolder({ displayName: 'SDK Test Folder', projectId })
  created.folderId = folder.id
  ok('createFolder()', `id=${folder.id}`)

  const updated = await ap.updateFolder(folder.id, { displayName: 'SDK Test Folder Updated' })
  ok('updateFolder()', `name=${updated.displayName}`)

  const fetched = await ap.getFolder(folder.id)
  ok('getFolder()', `id=${fetched.id}`)
} catch (e) { fail('Folders', e) }

// ── App Connections ───────────────────────────────────────────────────────────

console.log('\n[6] App Connections')
try {
  const res = await ap.listConnections({ limit: 3 })
  ok('listConnections()', `${res.data.length} connections`)

  if (res.data.length > 0) {
    const conn = await ap.getConnection(res.data[0].id)
    ok('getConnection()', `id=${conn.id} piece=${conn.pieceName}`)
  } else {
    skip('getConnection()', 'no connections available')
  }
} catch (e) { fail('App Connections', e) }

// ── Pieces ────────────────────────────────────────────────────────────────────

console.log('\n[7] Pieces')
try {
  const res = await ap.listPieces({ limit: 5 })
  ok('listPieces()', `${res.length} pieces, first=${res[0]?.displayName}`)
} catch (e) { fail('listPieces()', e) }

// ── Triggers ──────────────────────────────────────────────────────────────────

console.log('\n[8] Triggers')
try {
  const res = await ap.getTriggerRunStatus()
  const pieceCount = Object.keys(res.pieces ?? {}).length
  ok('getTriggerRunStatus()', `${pieceCount} pieces tracked`)
} catch (e) { fail('getTriggerRunStatus()', e) }

// ── Tables & Records ──────────────────────────────────────────────────────────

console.log('\n[9] Tables & Records')
try {
  const tables = await ap.listTables({ limit: 3 })
  ok('listTables()', `${tables.data.length} tables`)

  if (tables.data.length > 0) {
    const table = await ap.getTable(tables.data[0].id)
    ok('getTable()', `id=${table.id}`)

    const records = await ap.listRecords({ tableId: table.id, limit: 3 })
    ok('listRecords()', `${records.data.length} records`)
  } else {
    skip('getTable() / listRecords()', 'no tables available')
  }
} catch (e) { fail('Tables & Records', e) }

// ── MCP Servers ───────────────────────────────────────────────────────────────

console.log('\n[10] MCP Servers')
try {
  const flows = await ap.listFlows({ limit: 1 })
  const projectId = flows.data[0]?.projectId
  if (!projectId) throw new Error('No projectId available')

  const res = await ap.listMcpServers(projectId)
  ok('listMcpServers()', `${res.data.length} MCP servers`)

  if (res.data.length > 0) {
    const server = await ap.getMcpServer(res.data[0].id)
    ok('getMcpServer()', `id=${server.id}`)
  } else {
    skip('getMcpServer()', 'no MCP servers available')
  }
} catch (e) { fail('MCP Servers', e) }

// ── User Invitations ──────────────────────────────────────────────────────────

console.log('\n[11] User Invitations')
try {
  const res = await ap.listInvitations({ type: 'PLATFORM', limit: 3 })
  ok('listInvitations(type=PLATFORM)', `${res.data.length} invitations`)
} catch (e) { fail('listInvitations()', e) }

try {
  const res = await ap.listInvitations({ type: 'PROJECT', limit: 3 })
  ok('listInvitations(type=PROJECT)', `${res.data.length} invitations`)
} catch (e) { fail('listInvitations(type=PROJECT)', e) }

// ── Cleanup ───────────────────────────────────────────────────────────────────

console.log('\n[12] Cleanup')
if (created.folderId) {
  try {
    await ap.deleteFolder(created.folderId)
    ok(`deleteFolder(${created.folderId})`)
  } catch (e) { fail('deleteFolder()', e) }
}
if (created.flowId) {
  try {
    await ap.deleteFlow(created.flowId)
    ok(`deleteFlow(${created.flowId})`)
  } catch (e) { fail('deleteFlow()', e) }
}

// ── Summary ───────────────────────────────────────────────────────────────────

console.log(`\n${'─'.repeat(55)}`)
console.log(`  ${passed} passed  |  ${failed} failed  |  ${skipped} skipped`)
if (failed > 0) process.exit(1)
